libportal/settings.c

```diff
@@ -0,0 +1,191 @@
+ * License along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * SPDX-License-Identifier: LGPL-3.0-only
+ *
+ * Authors:
+ *    Hubert Figuière <hub@figuiere.net>
+ */
+
+#include "config.h"
+
+#include <libportal/portal.h>
+#include <libportal/settings.h>
+
+#include "portal-private.h"
+
+/**
+ * XdpSettings
+ *
+ * A representation of the settings exposed by the portal.
+ *
+ * The [class@Settings] object is used to access and observe the settings
+ * exposed by xdg-desktop-portal.
+ *
+ * It is obtained from [method@Portal.get_settings]. Call [method@Settings.read]
+ * to read a settings value. Connect to [signal@Settings::changed] to observe
+ * value changes.
+ */
+struct _XdpSettings {
+  GObject parent_instance;
+
+  XdpPortal *portal;
+  guint signal_id;
+
+};
+
+enum {
+  CHANGED,
+  LAST_SIGNAL
+};
+
+static guint signals[LAST_SIGNAL];
+
+G_DEFINE_TYPE (XdpSettings, xdp_settings, G_TYPE_OBJECT)
+
+static void
+xdp_settings_finalize (GObject *object)
+{
+  XdpSettings *settings = XDP_SETTINGS (object);
+
+  if (settings->signal_id)
+    g_dbus_connection_signal_unsubscribe (settings->portal->bus, settings->signal_id);
+
+  g_clear_object (&settings->portal);
+
+  G_OBJECT_CLASS (xdp_settings_parent_class)->finalize (object);
+}
+
+static void
+xdp_settings_class_init (XdpSettingsClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->finalize = xdp_settings_finalize;
+
+  /**
+   * XdpSettings::changed:
+   * @settings: the [class@Settings] object
+   * @namespace: the value namespace
+   * @key: the value key
+   * @value: the value
+   *
+   * Emitted when a setting value is changed externally.
+   */
+  signals[CHANGED] =
+    g_signal_new ("changed",
+                  G_TYPE_FROM_CLASS (object_class),
+                  G_SIGNAL_RUN_FIRST,
+                  0,
+                  NULL, NULL,
+                  NULL,
+                  G_TYPE_NONE, 3,
+                  G_TYPE_STRING,
+                  G_TYPE_STRING,
+                  G_TYPE_VARIANT);
+}
+
+static void
+xdp_settings_init (XdpSettings *settings)
+{
+}
+
+static void
+settings_changed (GDBusConnection *bus,
+		  const char *sender_name,
+		  const char *object_path,
+		  const char *interface_name,
+		  const char *signal_name,
+		  GVariant *parameters,
+		  gpointer data)
+{
+  const char *namespace = NULL;
+  const char *key = NULL;
+  g_autoptr(GVariant) value = NULL;
+  guint n_params;
+
+  XdpSettings *settings = data;
+
+  n_params = g_variant_n_children (parameters);
+  if (n_params != 3)
+    {
+      g_warning ("Incorrect number of parameters, expected 3, got %u", n_params);
+      return;
+    }
+
+  g_variant_get_child (parameters, 0, "&s", &namespace);
+  g_variant_get_child (parameters, 1, "&s", &key);
+  g_variant_get_child (parameters, 2, "v", &value);
+
+  g_signal_emit (settings, signals[CHANGED], 0, namespace, key, value);
+}
+
+/**
+ * xdp_settings_read:
```

Comment:
I'd love to see GSettings style wrappers around this:

`char *xdp_settings_read_string()`
`int xdp_settings_read_int()`

That can be a follow up though.

But in preparation for that this could be named `xdp_settings_read_value()` maybe?

Then we could have `xdp_settings_read()` take a GVariant format string like `g_settings_get()`.

Reply:
I think `xdp_settings_read_value` fine on its own but then `xdp_settings_read_all` is out of place. But `xdp_settings_read_all_values` is also kinda awkward? Thoughts?

Reply:
I have added these API. I kept the GCancellable an GError.